waku/common/ratelimit.nim

```diff
@@ -23,3 +23,28 @@ proc newTokenBucket*(setting: Option[RateLimitSetting]): Option[TokenBucket] =
     return none[TokenBucket]()
 
   return some(TokenBucket.new(volume, period))
+
+proc checkUsage*(
+    t: var Option[TokenBucket], proto: string, conn: Connection
+): bool {.raises: [Defect].} =
+  if t.isNone:
+    return true
+
+  let tokenBucket = t.get()
+  if not tokenBucket.tryConsume(1):
+    return false
+
+  return true
+
+template checkUsageLimit*(
+    t: var Option[TokenBucket],
+    proto: string,
+    conn: Connection,
+    bodyWithinLimit, bodyRejected: untyped,
```

Comment:
First time y encounter a template with two bodies. Want to make sure I read it correctly, as I find it a bit confusing.

First body goes goes after the `:` as usual and the second in a `do:` block?

Reply:
Exactly. I got the concept from nim std Table withValue template. https://nim-lang.org/docs/tables.html#withValue.t%2CTable%5BA%2CB%5D%2CA%2Cuntyped%2Cuntyped%2Cuntyped
as it perfectly suites the situation.

So first block is success scenario where request is within the limit. The `do` block will catch the over limit situation and allow to tailor the response buffer properly.

The issue was that different protocols have different response approaches (unfortunately) and needed to make it easy to define but still keep some consistent usage of the limit handling as it shall always happen before reading message into our buffer and decode it.